packages/react/src/host/RemoteRenderer.tsx

```diff
@@ -1,41 +1,55 @@
-import {memo} from 'react';
+import {cloneElement, memo, ReactElement, useMemo} from 'react';
 import {KIND_COMPONENT, KIND_TEXT, RemoteReceiver} from '@remote-ui/core';
 
 import type {Controller} from './controller';
 import {useAttached} from './hooks';
-import {RemoteText} from './RemoteText';
-import {RemoteComponent} from './RemoteComponent';
+import {renderText as defaultRenderText} from './RemoteText';
+import {renderComponent as defaultRenderComponent} from './RemoteComponent';
+import {RemoteRendererContext} from './context';
+import type {RemoteComponentProps, RemoteTextProps} from './types';
 
-interface Props {
+export interface RemoteRendererProps {
   receiver: RemoteReceiver;
   controller: Controller;
+  renderComponent?: (props: RemoteComponentProps) => ReactElement;
+  renderText?: (props: RemoteTextProps) => ReactElement;
 }
 
-export const RemoteRenderer = memo(({controller, receiver}: Props) => {
-  const {children} = useAttached(receiver, receiver.attached.root)!;
+export const RemoteRenderer = memo(
+  ({
+    controller,
+    receiver,
+    renderComponent = defaultRenderComponent,
+    renderText = defaultRenderText,
+  }: RemoteRendererProps) => {
+    const {children} = useAttached(receiver, receiver.attached.root)!;
+    const renderContextValue = useMemo(() => ({renderText, renderComponent}), [
```

Comment:
Does this need to be a separate concept from the `Controller`, which already has a very related job of mapping remote component names => host component implementations?

Reply:
Controller is meant for rendering component type. We also has Text Component as component type which will be rendered using text type. This renderText is for text type. 

Are you saying we should reserve a special Component in components list to render text type? This is what @vividviolet suggested before.

Reply:
Controller can be whatever we like it to be. Right now, it just gives the component implementation for a remote component, but I think it makes sense for `createController` to accept a second argument with the `renderText` stuff you've added here, and have those properties be exposed publicly such that we don't need a second piece of context.

Reply:
We could do it 👍